lib/iris/util.py

```diff
@@ -1507,6 +1507,98 @@ def demote_dim_coord_to_aux_coord(cube, name_or_coord):
+    """
+    Flatten the Aux coord and associated dimensions on the cube
+
+
+    * cube
+        An instance of :class:`iris.cube.Cube`
+
+    * name_or_coord:
+        Either
+
+        (a) An instance of :class:`iris.coords.DimCoord`
+
+        or
+
+        (b) the :attr:`standard_name`, :attr:`long_name`, or
+        :attr:`var_name` of an instance of an instance of
+        :class:`iris.coords.DimCoord`.
+
+    For example::
+
+        >>> print cube
+        rainfall_rate / (kg m-2 s-1)        (x: 5; y: 3)
+             Dimension coordinates:
+                  x                           x     -
+                  y                           -     x
+             Auxiliary coordinates:
+                  latitude                    x     x
+                  longitude                   x     x
+
+        >>> flatten_multidim_coord(cube, 'longitude')
+        >>> print cube
+        rainfall_rate / (kg m-2 s-1)        (--: 15)
+             Auxiliary coordinates:
+                  latitude                    x
+                  longitude                   x
+    """
+    import numpy as np
+    from iris.coords import DimCoord
+    from iris.cube import Cube
+
+    if isinstance(name_or_coord, six.string_types):
+        coord = cube.coord(name_or_coord)
+    elif isinstance(name_or_coord, iris.coords.Coord):
+        coord = name_or_coord
+    else:
+        # Don't know how to handle this type
+        msg = ("Don't know how to handle coordinate of type {}. "
+               "Ensure all coordinates are of type six.string_types or "
+               "iris.coords.Coord.")
+        msg = msg.format(type(name_or_coord))
+        raise TypeError(msg)
+
+    coord_dims = cube.coord_dims(coord)
+    other_dims = list(set(range(cube.ndim)) - set(coord_dims))
+
+    # Get all the coords which span the dimensions to be flattened
+    coords_to_flatten = cube.coords(dimensions=coord_dims)
+    # And any other coords which span any one of the dimensions
+    for d in coord_dims:
+        coords_to_flatten.extend(cube.coords(dimensions=d))
+
+    other_dim_coords = [(c, i) for i, c in
+                        enumerate(cube.coords(dim_coords=True))
+                        if c not in coords_to_flatten]
+    other_aux_coords = [(c, cube.coord_dims(c)) for c in
+                        cube.coords(dim_coords=False)
+                        if c not in coords_to_flatten]
+
+    shape_array = np.asarray(cube.shape)
+    new_shape = np.hstack([shape_array[other_dims],
+                           np.product(shape_array[list(coord_dims)])])
+
+    new_data = cube.data.reshape(new_shape)
+    new_aux_coords = other_aux_coords
+    for c in coords_to_flatten:
+        # Only flatten aux coords (we will drop DimCoords)
+        if not isinstance(c, DimCoord):
+            # The new (flat) AuxCoords are always the last dimension
+            new_aux_coords.append((c.copy(c.points.flat,
+                                          c.bounds.reshape(-1, 2)),
```

Comment:
I got an error here because `bounds` is None.

Reply:
I'm also not sure if this is robust to the cases where AuxCoord dimensions are not necessarily in the order you expect (#2606).

Reply:
I'll fix that